src/Integrals.jl

```diff
@@ -147,5 +147,107 @@ function __solvebp_call(prob::IntegralProblem, alg::VEGAS, sensealg, lb, ub, p;
     SciMLBase.build_solution(prob, alg, val, err, chi = chi, retcode = ReturnCode.Success)
 end
 
-export QuadGKJL, HCubatureJL, VEGAS, GaussLegendre
+is_sampled_problem(prob::IntegralProblem) = prob.f isa AbstractArray
+import SciMLBase.IntegralProblem # this is type piracy, and belongs in SciMLBase
+function IntegralProblem(y::AbstractArray, lb, ub, args...; kwargs...)
+    IntegralProblem{false}(y, lb, ub, args...; kwargs...)
+end
+function construct_grid(prob, alg, lb, ub, dim)
+    x = alg.spec
+    @assert length(ub) == length(lb) == 1 "Multidimensional integration is not supported with the Trapezoidal method"
+    if x isa Integer
+        grid = range(lb[1], ub[1], length=x)
+    else 
+        grid = x
+        @assert ndims(grid) == 1 "Multidimensional integration is not supported with the Trapezoidal method"
+    end
+
+    @assert lb[1] ≈ grid[begin] "Lower bound in `IntegralProblem` must coincide with that of the grid"
+    @assert ub[1] ≈ grid[end] "Upper bound in `IntegralProblem` must coincide with that of the grid"
+    if is_sampled_problem(prob)
+        @assert size(prob.f, dim) == length(grid) "Integrand and grid must be of equal length along the integrated dimension"
+        @assert axes(prob.f, dim) == axes(grid,1) "Grid and integrand array must use same indexing along integrated dimension" 
+    end
+    return grid
+end
+
+@inline myselectdim(y::AbstractArray{T,dims}, d, i) where {T,dims} = selectdim(y, d, i)
+@inline myselectdim(y::AbstractArray{T,1}, _, i) where {T} = @inbounds y[i]
+
+@inline dimension(::Val{D}) where D = D
+function __solvebp_call(prob::IntegralProblem, alg::Trapezoidal{S, D}, sensealg, lb, ub, p; kwargs...) where {S,D}
+    # since all AbstractRange types are equidistant by design, we can rely on that
+    @assert prob.batch == 0
+    # using `Val`s for dimensionality is required to make `selectdim` not allocate
+    dim = dimension(D) 
+    p = p
+    if is_sampled_problem(prob)
+        @assert alg.spec isa AbstractArray "For pre-sampled problems where the integrand is an array, the integration grid must also be specified by an array."
+    end
+
+    grid = construct_grid(prob, alg, lb, ub, dim)
+    
+    err = Inf64
+    if is_sampled_problem(prob)
+        data = prob.f
+        # inlining is required in order to not allocate
+        @inline function integrand(i) 
```

Comment:
```suggestion
        integrand = function (i) 
```

It should be an anonymous function if in an if statement